File: PSet1/credit.c
#include <stdio.h>
#include <cs50.h>

long long getCard()
{
	printf("Please provide your credit card number: ");
	return GetLongLong();
}

int firstDigit(long long cardNum)
{
	while (cardNum > 100)
		cardNum = cardNum / 10;

	return (int) cardNum;
}


bool checksum(long long cardNum)
{
		int sum1 = 0;
		int sum2 = 0;

		while (cardNum > 0)
    	{
    		//step 1
    		sum1 += cardNum % 10;
    		cardNum = cardNum / 10;

    		// step2
    		int toMultiply = (cardNum % 10) * 2;

    		if (toMultiply >= 10)
    		{
    			toMultiply = (toMultiply % 10) + (toMultiply / 10);
    		}
    		sum2 += toMultiply;
    		cardNum = cardNum/10;

    	};

    	int result = sum1 + sum2;
    	if (result % 10 == 0)
    		return true;
    	else
    		return false;
}

void printCardType(int firstDigit)
{
	if (firstDigit < 50 && firstDigit >= 40)
	{
		// Visa
		printf("VISA\n");
	}
	else if (firstDigit <= 55 && firstDigit >= 51 )
	{
		// Amex
		printf("MASTERCARD\n");
	}
	else if (firstDigit == 34 || firstDigit == 37)
	{
		// Mastercard
		printf("AMEX\n");
	}
	else
		printf("INVALID\n");
}

int main()
{
	long long cardNum = getCard();

	if (checksum(cardNum) == true)
		printCardType(firstDigit(cardNum));
	else printf("INVALID\n");
}

File: Pset6/python/credit.py
#!/usr/bin/python3

def first_digit(card_num):
    while card_num < 100:
        card_num = card_num / 10

    print("card_num = " + card_num)
    return card_num

def checksum(card_num):
    sum_1 = 0
    sum_2 = 0

    while card_num > 0:
        #step 1
        sum_1 += card_num % 10
        card_num = int(card_num / 10)
        
        # step 2
        to_multiply = (card_num % 10) * 2

        if to_multiply >= 10:
            to_multiply = (to_multiply % 10) + int(to_multiply / 10)
        
        sum_2 += to_multiply
        card_num = int(card_num / 10)

    print(sum_1)
    print(sum_2)
    result = sum_1 + sum_2
    print(type(result))
    #print("result = " + str(result))
    if (result % 10) == 0:
            print("TRUE")
            return True
    else: 
            print("FALSE")
            return False

def print_card_type(first_digit):
    # if 40 >= first_digit > 50:
    if first_digit in range (40,51):
        print("VISA")
    elif first_digit in range (51,56):
        print("MASTERCARD")
    elif first_digit in (34, 37):
        print("VISA")
    else: print("INVALID")

if __name__ == '__main__':
    card_num  = input("Please provide your credit card number : ")

    if checksum(int(card_num)):
        print_card_type(first_digit(int(card_num)))
    else: print('INVALID')
